p11-kit/rpc-server.c

```diff
@@ -1834,6 +1878,416 @@ rpc_C_GenerateRandom (CK_X_FUNCTION_LIST *self,
+	CK_FLAGS flags;
+
+	BEGIN_CALL (DecryptMessageNext);
+		IN_ULONG (session);
+		IN_BYTE_ARRAY (parameter, parameter_len);
+		IN_BYTE_ARRAY (ciphertext_part, ciphertext_part_len);
+		IN_BYTE_BUFFER (plaintext_part, plaintext_part_len);
+		IN_ULONG (flags);
+	PROCESS_CALL ((self, session, (void *)parameter, parameter_len, ciphertext_part, ciphertext_part_len,
+	               plaintext_part, &plaintext_part_len, flags));
+		OUT_BYTE_ARRAY (plaintext_part, plaintext_part_len);
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_MessageDecryptFinal (CK_X_FUNCTION_LIST *self,
+                           p11_rpc_message *msg)
+{
+	CK_SESSION_HANDLE session;
+
+	BEGIN_CALL (MessageDecryptFinal);
+		IN_ULONG (session);
+	PROCESS_CALL ((self, session));
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_MessageSignInit (CK_X_FUNCTION_LIST *self,
+                       p11_rpc_message *msg)
+{
+	CK_SESSION_HANDLE session;
+	CK_MECHANISM mechanism_;
+	CK_MECHANISM_PTR mechanism = &mechanism_;
+	CK_OBJECT_HANDLE key;
+
+	BEGIN_CALL (MessageSignInit);
+		IN_ULONG (session);
+		IN_MECHANISM (mechanism);
+		IN_ULONG (key);
+	PROCESS_CALL ((self, session, mechanism, key));
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_SignMessage (CK_X_FUNCTION_LIST *self,
+                   p11_rpc_message *msg)
+{
+	CK_SESSION_HANDLE session;
+	CK_BYTE_PTR parameter;
+	CK_ULONG parameter_len;
+	CK_BYTE_PTR data;
+	CK_ULONG data_len;
+	CK_BYTE_PTR signature;
+	CK_ULONG signature_len;
+
+	BEGIN_CALL (SignMessage);
+		IN_ULONG (session);
+		IN_BYTE_ARRAY (parameter, parameter_len);
+		IN_BYTE_ARRAY (data, data_len);
+		IN_BYTE_BUFFER (signature, signature_len);
+	PROCESS_CALL ((self, session, (void *)parameter, parameter_len, data, data_len,
+	               signature, &signature_len));
+		OUT_BYTE_ARRAY (signature, signature_len);
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_SignMessageBegin (CK_X_FUNCTION_LIST *self,
+                        p11_rpc_message *msg)
+{
+	CK_SESSION_HANDLE session;
+	CK_BYTE_PTR parameter;
+	CK_ULONG parameter_len;
+
+	BEGIN_CALL (SignMessageBegin);
+		IN_ULONG (session);
+		IN_BYTE_ARRAY (parameter, parameter_len);
+	PROCESS_CALL ((self, session, (void *)parameter, parameter_len));
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_SignMessageNext (CK_X_FUNCTION_LIST *self,
+                       p11_rpc_message *msg)
+{
+	CK_SESSION_HANDLE session;
+	CK_BYTE_PTR parameter;
+	CK_ULONG parameter_len;
+	CK_BYTE_PTR data;
+	CK_ULONG data_len;
+	CK_BYTE_PTR signature;
+	CK_ULONG signature_len_;
+	CK_ULONG_PTR signature_len = &signature_len_;
+
+	BEGIN_CALL (SignMessageNext);
+		IN_ULONG (session);
+		IN_BYTE_ARRAY (parameter, parameter_len);
+		IN_BYTE_ARRAY (data, data_len);
+		_ret = proto_read_byte_buffer_null (msg, &signature, &signature_len);
+		if (_ret != CKR_OK)
+			goto _cleanup;
+	PROCESS_CALL ((self, session, (void *)parameter, parameter_len, data, data_len,
+	               signature, signature_len));
+		OUT_BYTE_ARRAY (signature, signature_len ? *signature_len : 0);
+	END_CALL;
```

Comment:
Sorry, what I really meant was to change the protocol [signature](https://github.com/p11-glue/p11-kit/pull/463/files#diff-4c71e18b8f308838a79477ed5669813f625782daf55b2dd28714fad582afa091R245) of C_SignMessageNext from "uayayfy" to "uayayyfy" (note the extra "y" before "fy"). Then you could write this part something like:

```suggestion
	BEGIN_CALL (SignMessageNext);
		IN_ULONG (session);
		IN_BYTE_ARRAY (parameter, parameter_len);
		IN_BYTE_ARRAY (data, data_len);
		IN_BYTE (signature_expected);
		IN_BYTE_BUFFER (signature, signature_len);
	PROCESS_CALL ((self, session, (void *)parameter, parameter_len, data, data_len,
	               signature_expected ? signature : NULL, signature_expected ? signature_len : NULL));
		OUT_BYTE_ARRAY (signature, signature_expected ? *signature_len : 0);
	END_CALL;
```

Reply:
could you take a look at this part again? I think rpc-server should be good now. Dont know about rpc-client